### ISSUE TITLE
Implement ViCo: Detail-Preserving Visual Condition for Personalized Text-to-Image Generation for diffusers

### ISSUE THREAD

**Issue by okaris** [huggingface/diffusers#3719]
I would like to use new advancements for personalisation

To the best of my knowledge [ViCo](https://github.com/haoosz/ViCo) is the first available open source implementation of similar ideas that also use an image encoding for conditioning diffusion models.

https://github.com/haoosz/ViCo

**Comment by sakshamjindal** [comment]
There is another which seems promising which use image-aligned text embeddings and does a zero-shot "dreambooth". https://dxli94.github.io/BLIP-Diffusion-website/

**Comment by okaris** [comment]
@sakshamjindal Thank you for sharing that repo. If i'm reading correctly it's still not zero-shot and requires some fine-tuning too.

**Comment by patrickvonplaten** [comment]
This looks indeed quite interesting! Would be more than happy to review a PR - sadly don't have time at the moment to look into it myself. Also cc @sayakpaul here

**Comment by sayakpaul** [comment]
Sure, we'd be happy to accept a PR and guide the contributor along the way :)

**Comment by okaris** [comment]
@patrickvonplaten @sayakpaul I'm taking a first stab at this myself. After reviewing the main branch and a few related issues, here is how I decided to approach this:

* ViCo needs 2 unet passes, one with the conditioning image and one with the noisy latent. 
* Attention maps need to be extracted from the first pass and fed into the new attention mechanism in the 2nd pass.
* To avoid loading 2 copies of unet I use the same unet with a new custom attention processor
* This processor saves the attention maps to an attention store (like Attend and Excite)
* First pass of the unet writes the maps, second pass of the unet reads the maps, store is reset after second pass
* For the second pass the new attention processor introduces a new attention block (therefore new layers, like Custom Diffusion)

I'll test this naive approach first before abstracting a new pipeline to utilise this new attention processor.

Please let me know if this is the right approach.

Thank you

**Comment by sayakpaul** [comment]
Like the approach especially your plan on avoiding loading two instances of the UNet.

**Comment by okaris** [comment]
@sayakpaul @patrickvonplaten I have succeeded with a naive implementation using a custom attention processor. If I wrap this up in a new "pipeline" it should be fine but I believe an implementation at the block level would be better.

Is there an easy way (like how we replace attention processors) to replace all the cross attention blocks used in a pipeline  

"CrossAttnDownBlock2D",
"CrossAttnDownBlock2D",
"CrossAttnDownBlock2D",
"DownBlock2D",
"UpBlock2D",
"CrossAttnUpBlock2D",
"CrossAttnUpBlock2D",
"CrossAttnUpBlock2D"

without rewriting the whole unet?

Thanks!

**Comment by sayakpaul** [comment]
I don't think we can support that just yet. We usually introduce function / class arguments to facilitate those changes. This is the approach we followed for ControlNet. (Cc: @williamberman @takuma104).

**Comment by okaris** [comment]
Circling back here, I wasn't able to find a way to implement it without directly changing the blocks. The only stop gap solution is to edit the config before loading the model which is not ideal. Any ideas?

**Comment by sayakpaul** [comment]
Hi @okaris!

Could you maybe a post some small snippets to describe more concretely what you're looking for to aid the implementation? We can then maybe provide something more concrete for you.

**Comment by okaris** [comment]
The ViCo Unet accepts noisey latents, visual condition latent(s), text condition embeds

These are passed on the every transformer block

```
sample, sample_image, loss_reg = upsample_block(
                    hidden_states=sample,
                    hidden_states_image=sample_image,
                    temb=emb,
                    res_hidden_states_tuple=res_samples,
                    res_hidden_states_image_tuple=res_samples_image,
                    encoder_hidden_states=encoder_hidden_states,
                    cross_attention_kwargs=cross_attention_kwargs,
                    upsample_size=upsample_size,
                    attention_mask=attention_mask,
                    encoder_attention_mask=encoder_attention_mask,
                    ph_pos=ph_pos,
                )
```

noisy latents and visual condition latents are passed through each block. visual features are extracted from the visual condition latents and attention scores are extracted

```
hidden_states = resnet(hidden_states, temb)
                hidden_states_image = resnet(hidden_states_image, temb)

                hidden_states, hidden_states_image, loss_reg = attn(
                    hidden_states,
                    hidden_states_image,
                    encoder_hidden_states=encoder_hidden_states,
                    cross_attention_kwargs=cross_attention_kwargs,
                    attention_mask=attention_mask,
                    encoder_attention_mask=encoder_attention_mask,
                    return_dict=False,
                    ph_pos=ph_pos,
                )
```

in the cross attention transformer, if visual condition is enabled (it is only enable for **some** of the **up blocks**) the attentions for latents and visual condition are passed through a 3rd attention transformer to generate the final attention. this is also where the mask generated from the attention scores of visual condition at the subject token position are built, as well as the loss regularization between the attention scores of the subject token and eos token

```
  hidden_states = block(
                hidden_states,
                attention_mask=attention_mask,
                encoder_hidden_states=encoder_hidden_states,
                encoder_attention_mask=encoder_attention_mask,
                timestep=timestep,
                cross_attention_kwargs=cross_attention_kwargs,
                class_labels=class_labels,
            )

            hidden_states_image, attn = block(
                    hidden_states_image,
                    attention_mask=attention_mask,
                    encoder_hidden_states=encoder_hidden_states,
                    encoder_attention_mask=encoder_attention_mask,
                    timestep=timestep,
                    cross_attention_kwargs=cross_attention_kwargs,
                    class_labels=class_labels,
                    return_attn=True,
                )
        
                loss_reg = F.mse_loss( attn_subject, attn_eos )

                mask = build_mask(attn_scores) #omitted

                hidden_states = self.image_cross_attn(
                    hidden_states, 
                    encoder_hidden_states=hidden_states_image,
                    attention_mask=mask,
                    timestep=timestep,
                    cross_attention_kwargs=cross_attention_kwargs,
                    class_labels=class_labels,
                    )
```

finally here is the attention processor which also returns attention scores when it's needed, and when it's not uses the regular attention processor

```
 if return_attn:
            attention_scores = attn.get_attention_scores(query, key, attention_mask)
            hidden_states = torch.bmm(attention_scores, value)
        else:
            hidden_states = F.scaled_dot_product_attention(
                query, key, value, attn_mask=attention_mask, dropout_p=0.0, is_causal=False
            )
```

I can see some patterns that resemble this for Kandinsky 2.0 but not exactly

I had to modify the following classes:
UNet2DConditionModel
Transformer2DModel
BasicTransformerBlock
CrossAttnDownBlock2D
CrossAttnDownBlock2D
CrossAttnDownBlock2D
DownBlock2D
UpBlock2D
CrossAttnUpBlock2D
CrossAttnUpBlock2D
CrossAttnUpBlock2D
AttnProcessor

to achieve the exact results from the paper

**Comment by sayakpaul** [comment]
Thanks so much for being rigorous here and for sharing this!

I think we can condition the above blocks with suitable arguments. This way we don't have to rewrite those classes and we can promote reusability. Is there any difficulty you're facing in doing so?

**Comment by okaris** [comment]
Thank you. Let me make some adjustments to handle when the image condition is not passed. The only thing I didn’t figure out is how I can pass an argument when using from_pretrained so the new layers are not initialised

**Comment by yoshibenjo** [comment]
Any updates here? cc: @okaris

**Comment by sayakpaul** [comment]
Closing this due to lack of activities. Feel free to reopen.
